src/corecel/sys/KernelRegistry.hh

```diff
@@ -0,0 +1,126 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file corecel/sys/KernelRegistry.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <atomic>
+#include <iosfwd>
+#include <memory>
+#include <mutex>
+#include <string>
+#include <vector>
+
+#include "corecel/Assert.hh"
+#include "corecel/OpaqueId.hh"
+
+#include "KernelAttributes.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+struct KernelProfiling
+{
+    //!< Number of times launched
+    std::atomic<int> num_launches{0};
+    //!< Number of threads integrated over all launches
+    std::atomic<long long> accum_threads{0};
+
+    // Increment atomic counters given the number of threads
+    inline void log_launch(int num_threads);
+};
+
+//---------------------------------------------------------------------------//
+struct KernelMetadata
+{
+    std::string      name;
+    KernelAttributes attributes;
+    KernelProfiling  profiling;
+};
+
+//! Ordered identifiers for registered kernels
+using KernelId = OpaqueId<KernelMetadata>;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Keep track of kernels and launches.
+ *
+ * This class has a thread-safe *insert* method because it's meant to be shared
+ * across multiple threads when running. Since accessing by ID is expected to
+ * be done well after all insertions have been completed, the \c size and \c at
+ * accessors are *not* thread safe.
+ */
+class KernelRegistry
+{
+  public:
+    //!@{
+    //! \name Type aliases
+    using value_type      = KernelMetadata;
+    using const_reference = const KernelMetadata&;
+    using key_type        = KernelId;
+    using size_type       = KernelId::size_type;
+    //!@}
+
+  public:
+    // Whether profiling metrics (launch count, max threads) are collected
+    static bool profiling();
+
+    // Construct without any data
+    KernelRegistry() = default;
+
+    // Register a kernel and return optional reference to profiling info
```

Comment:
You may want to note here (or somewhere more adequate) that the underlying containers allows multiple entries with the same name (behave like a mutlimap not like map).  This is important because if the containers did not allow multiple entries it could either fail to insert or worse returns an existing entry .. and then breaking the thread safety.  (i.e. code is currently okay but a later "improvement" could easily break it).